src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,192 @@
+    ///
+    virtual CHIP_ERROR Complete(CHIP_ERROR error) = 0;
+};
+
+/// Enforces that once acquired, Complete will be called on the underlying writer
+class AutoCompleteInvokeResponder
+{
+public:
+    // non-copyable: once you have a handle, keep it
+    AutoCompleteInvokeResponder(const AutoCompleteInvokeResponder &)             = delete;
+    AutoCompleteInvokeResponder & operator=(const AutoCompleteInvokeResponder &) = delete;
+
+    AutoCompleteInvokeResponder(InvokeResponder * writer) : mWriter(writer) {}
+    ~AutoCompleteInvokeResponder()
+    {
+        if (!mCompleted)
+        {
+            mWriter->Complete(CHIP_IM_GLOBAL_STATUS(Failure));
+        }
+    }
+
+    /// Direct access to reply encoding.
+    ///
+    /// Use this only in conjunction with the other Raw* calls
+    DataModel::WrappedStructEncoder & RawReplyEncoder(CommandId replyCommandId) { return mWriter->ReplyEncoder(replyCommandId); }
+
+    /// Direct access to flushing replies
+    ///
+    /// Use this only in conjunction with the other Raw* calls
+    CHIP_ERROR RawFlushPendingReplies() { return mWriter->FlushPendingReplies(); }
+
+    /// Call "Complete" without the automatic retries.
+    ///
+    /// Use this in conjunction with the other Raw* calls
+    CHIP_ERROR RawComplete(CHIP_ERROR error) { return mWriter->Complete(error); }
+
+    /// Complete the given command.
+    ///
+    /// Automatically handles retries for sending.
+    CHIP_ERROR Complete(CHIP_ERROR error)
+    {
+        mCompleted     = true;
+        CHIP_ERROR err = mWriter->Complete(error);
+
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingReplies());
+        return mWriter->Complete(error);
+
+        return err;
+    }
+
+    /// Sends the specified data structure as a response
+    ///
+    /// This version of the send has built-in RETRY and handles
+    /// Flush/Complete automatically.
+    ///
+    template <typename ReplyData>
+    CHIP_ERROR Send(const ReplyData & data)
+    {
+        CHIP_ERROR err = data.Encode(ReplyEncoder(ReplyData::GetCommandId()));
+        LogErrorOnFailure(err);
+        err = mWriter->Complete(err);
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingReplies());
+        err = data.Encode(ReplyEncoder(ReplyData::GetCommandId()));
+        LogErrorOnFailure(err);
+        return mWriter->Complete(err);
+    }
+
+private:
+    InvokeResponder * mWriter;
+    bool mCompleted = false;
+};
+
+enum ReplyAsyncFlags
+{
+    // Some commmands that are expensive to process (e.g. crypto)
+    // should ack right away to improve reliability and reduce
+    // needless retries.
+    kFlushAckRightAway = 0x0001,
```

Comment:
In one of my recent PRs Boris wanted to move away from FlushAckRightAway to HandlingSlowCommand. That way apps don't care about underlying transport mechanism and simply provide a hint that they are going to take some time to handle slow command

Reply:
My understanding is that from (Application + Data Model and Interaction Model) we should not need to know anything about the transport layer (exchange). The former should only care about indicating it is processing a command that will take and provide that hint to later layers. The later transport layers can translate that hint of processing a slow command to flushing a stand alone ack

Reply:
Updated the name.